### PR TITLE
chore(cli): enable Windows CLI E2E tests on GitHub Actions

### PR DESCRIPTION
**Pull request by byCedric**
# Why

Split off from #33204, stacked on #33401 ([for the `EXPO_E2E_TEMP_DIR` workaround](https://github.com/expo/expo/pull/33387#discussion_r1871455847))

This enables running CLI E2E tests on Windows, allowing it to fail until we've merged all fixes required for Windows.

> [!IMPORTANT]
> We should merge this _after_ all Windows fixes have been merged, to ensure we cleared all current Windows pathing issues surfaced from the Jest E2E tests.
> - https://github.com/expo/expo/pull/33401
> - https://github.com/expo/expo/pull/33408
> - https://github.com/expo/expo/pull/33397
> - https://github.com/expo/expo/pull/33396
> - https://github.com/expo/expo/pull/33395
> - https://github.com/expo/expo/pull/33394
> - https://github.com/expo/expo/pull/33393

# How

- Modified **.github/workflows/cli.yml** to include Windows (and allow it to fail).

# Test Plan

- See GitHub Action

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by Kudo — APPROVED** [review]

File: .github/workflows/cli.yml [hunk 1]
```diff
@@ -42,52 +42,124 @@ concurrency:
...
+
+      - name: 🔎 Type Check CLI (Windows)
         working-directory: packages/@expo/cli
+        # TODO(cedric): expo module scripts is incompatible with Windows, forcing us to manually set up these steps instead of `yarn typecheck`
+        run: yarn tsc --noEmit
 
+      - name: 🧪 E2E Test CLI (Windows)
+        working-directory: packages/@expo/cli
+        # NOTE(cedric): Windows is significantly slower and suffers from IO congestion when running more than 2 workers
+        run: yarn test:e2e --max-workers 2
```
Comment (byCedric) [inline .github/workflows/cli.yml:161]: I've also embedded this into #33401, where it switches tests to run in serial mode when GHA is running in debug mode. Probably can drop the `--max-workers 2` here.

Comment (byCedric) [reply .github/workflows/cli.yml:161]: Another requirement for this is that we have to modify the temporary system folder on Windows/GHA to be on the same disk as the repository checkout. The issue there is that Metro's FileMap only "scopes" a single disk on Windows [due to its design](https://github.com/facebook/metro/blob/main/packages/metro-file-map/src/lib/TreeFS.js#L96). Meaning that on Windows, it only scopes to `C:\\...` or `D:\\...`.

GHA uses both `C:` (system temp dir) and `D:` (repository checkout) drives, causing `SHA1 not calculated` errors when running Expo CLI from source. I've tested this in: https://github.com/expo/expo/pull/33204/commits/cbe4ca09ccccc28131b8c2a892813394be8bbc91 - which finally got [Windows to pass all E2E tests](https://github.com/expo/expo/actions/runs/12168114455/job/33938213550)

File: .github/workflows/cli.yml [hunk 2]
```diff
@@ -42,52 +42,125 @@ concurrency:
...
+      - name: 🔎 Type Check CLI
+        working-directory: packages/@expo/cli
+        run: yarn typecheck
+
+      - name: 🧪 E2E Test CLI
+        working-directory: packages/@expo/cli
+        run: yarn test:e2e --shard ${{ matrix.shard }}/${{ strategy.job-total }}
+
+  e2e-windows:
+    runs-on: windows-2022
```
Comment (byCedric) [inline .github/workflows/cli.yml:101]: TODO: add sharding once the E2E tests are passing on Windows, to speed things up without unnecessary spamming with error messages.

File: .github/workflows/cli.yml [hunk 3]
```diff
@@ -42,52 +42,125 @@ concurrency:
...
+    continue-on-error: true
```
Comment (byCedric) [inline .github/workflows/cli.yml:102]: TODO
- Add sharding once the E2E tests are passing on Windows, to speed things up without unnecessary spamming with error messages.
- Remove the `continue-on-error: true` once the test are passing on Windows

File: .github/workflows/cli.yml [hunk 4]
```diff
@@ -42,52 +42,125 @@ concurrency:
   cancel-in-progress: true
 
 jobs:
-  test:
+  e2e-ubuntu:
```
Comment (byCedric) [inline .github/workflows/cli.yml:45]:
```suggestion
  jest-ubuntu:
```

Maybe this makes more sense, as playwright is also running E2E tests.

File: .github/workflows/cli.yml [hunk 5]
```diff
@@ -42,52 +42,125 @@ concurrency:
...
 
+      - name: 🔎 Type Check CLI
+        working-directory: packages/@expo/cli
+        run: yarn typecheck
+
+      - name: 🧪 E2E Test CLI
+        working-directory: packages/@expo/cli
+        run: yarn test:e2e --shard ${{ matrix.shard }}/${{ strategy.job-total }}
+
+  e2e-windows:
```
Comment (byCedric) [inline .github/workflows/cli.yml:100]:
```suggestion
  jest-windows:
```

File: .github/workflows/cli.yml [hunk 6]
```diff
@@ -42,52 +42,125 @@ concurrency:
...
-        run: yarn test:e2e --shard ${{ matrix.shard }}/${{ strategy.job-total }}
         working-directory: packages/@expo/cli
+        # TODO(cedric): expo module scripts is incompatible with Windows, forcing us to manually set up these steps instead of `yarn typecheck`
+        run: yarn tsc --noEmit
 
+      - name: 🧪 E2E Test CLI
+        working-directory: packages/@expo/cli
+        run: yarn test:e2e
+  
   web:
```
Comment (byCedric) [inline .github/workflows/cli.yml:164]:
```suggestion
  playwright-ubuntu:
```

**Review by byCedric — COMMENTED** [review]
Maybe this makes more sense, as playwright is also running E2E tests.

File: .github/workflows/cli.yml (hunk 1 above)
Comment (byCedric) [reply .github/workflows/cli.yml]: Both done!

**Comment by byCedric** [comment]
Ubuntu E2E failure is due to the test timing out and caused by IO congestion, causing the 2x export to take way longer than expected. We could increase the timeout, or limit the workers on ubuntu (and try to speed up using more shards).

Adding max workers has both pro and cons. The tests will execute slower, but the tests will also execute more accurately - where the full power of the machine is available for each test (instead of shared across all sharded tests).

File: .github/workflows/cli.yml (hunk 3 above)
Comment (byCedric) [inline .github/workflows/cli.yml:102]:
```suggestion
    strategy:
      fail-fast: false
      matrix:
        shard: [1, 2, 3, 4]
```

File: .github/workflows/cli.yml [hunk 7]
```diff
@@ -42,52 +42,125 @@ concurrency:
...
-        working-directory: packages/@expo/cli
-      - name: E2E Test CLI
-        run: yarn test:e2e --shard ${{ matrix.shard }}/${{ strategy.job-total }}
         working-directory: packages/@expo/cli
+        # TODO(cedric): expo module scripts is incompatible with Windows, forcing us to manually set up these steps instead of `yarn typecheck`
+        run: yarn tsc --noEmit
 
+      - name: 🧪 E2E Test CLI
+        working-directory: packages/@expo/cli
+        run: yarn test:e2e
```
Comment (byCedric) [inline .github/workflows/cli.yml:162]:
```suggestion
        run: yarn test:e2e ${{ matrix.shard }}/${{ strategy.job-total }}
```

**Review by byCedric — COMMENTED** [review]
TODO
- Add sharding once the E2E tests are passing on Windows, to speed things up without unnecessary spamming with error messages.
- Remove the `continue-on-error: true` once the test are passing on Windows

**Review by byCedric — COMMENTED** [review]
-

**Review by brentvatne — APPROVED** [review]
thanks for putting so much work into this @byCedric! looking forward to #33401 landing so we can start getting these tests running

**Pull request merged**